File: ch08/bucket_sort.py

import random
import math

def insertion_sort(v):
  n = len(v)
  if n > 0:
    for i in range(1, n):
      p = i - 1
      q = i
      while p >= 0 and v[p] > v[q]:
        exchange(v, p, q)
        p -= 1
        q -= 1

def bucket_sort(v):
  n = len(v)
  buk = []
  for i in range(0, n):
    buk.append([])

  for element in v:
    index = math.floor(n * element)
    buk[index].append(element)

  for B in buk:
    insertion_sort(B)

  k = 0
  for B in buk:
    for element in B:
      v[k] = element
      k += 1

#---------------------------------------

def exchange(v, i, j):
  temp = v[i]
  v[i] = v[j]
  v[j] = temp

def mkRandomList(size, n_min, n_max):
  bound = n_max - n_min
  v = []
  for i in range(0, size):
    v.append(random.randint(0, bound) + n_min)
  return v

def mkRandomList(size):
  v = []
  for i in range(0, size):
    v.append(random.random())
  return v

def printList(v):
  print('(List: #', end='')
  print(len(v), end='')
  print(' [', end='')
  for i in range(0, len(v)):
    print(v[i], end='')
    print(' ', end='')
  print('])')

A = mkRandomList(100)
printList(A)
bucket_sort(A)
printList(A)

File: ch08/bucket_sort.js

const insertion_sort = function(v) {
    let n = v.length;
    if (n > 0) {
        for (let i = 1; i != n; i += 1) {
            for (let p = i - 1, q = i;
                     p >= 0 && v[p] > v[q]; p -= 1, q -= 1) {
                exchange(v, p, q);
            }
        }
    }
}

const bucket_sort = function(v) {
    let bucket = new Array();
    let n = v.length;
    for (let i = 0; i != n; i += 1) {
        bucket[i] = new Array();
    }

    for (element of v) {
        let index = Math.floor(n * element);
        bucket[index].push(element);
    }

    for (b of bucket) {
        insertion_sort(b);
    }

    let k = 0;
    for (b of bucket) {
        for (element of b) {
            v[k] = element;
            k += 1;
        }
    }
}

//--------------------------------------

const exchange = function(v, i, j) {
    let temp = v[i];
    v[i] = v[j];
    v[j] = temp;
}

const randomInt = function(min, max) {
    return Math.floor(Math.random() * (max - min) + min);
}

const mkIntArray = function(size, min, max) {
    let v = new Array();
    for (let i = 0; i < size; i += 1) {
        v.push(randomInt(min, max));
    }
    return v;
}

const mkFloatArray = function(size) {
    let v = new Array();
    for (let i = 0; i < size; i += 1) {
        v.push(Math.random());
    }
    return v;
}

const printArray = function(v) {
    process.stdout.write("(Array #" + v.length + " [")
    for (element of v) {
        process.stdout.write(element + " ");
    }
    process.stdout.write("])\n");
}

let A = mkFloatArray(100);
printArray(A);
bucket_sort(A);
printArray(A);
